File: codesite/python_problems/static/python_problems/linked-list-utils.js
// export { ListNode, buildLinkedList, getLinkedListValues, areLinkedListsEqueal }


/**
 * Represents a node in a singly-linked list.
 * @class
 * @param {number|null} [val=null]
 * @param {ListNode|null} [next=null]
 */
class ListNode {
   constructor(val = null, next = null) {
      this.val = val;
      this.next = next;
   }
}


/**
 * Converts an array of values into a linked list
 * @param {number[]} numbers
 * @returns {ListNode}
 */
const buildLinkedList = (numbers, { cyclePosition = null } = {}) => {
   let node = new ListNode();
   const anchor = node;
   let hasCycle = false;
   let cycleNode;

   for (let position = 0; position < numbers.length; position++) {
      const number = numbers[position];
      node.next = new ListNode(number);
      node = node.next;

      if (position === cyclePosition) {
         cycleNode = node;
         hasCycle = true;
      }
   }
   if (hasCycle) {
      node.next = cycleNode;
   }
   return anchor.next
}


/**
 * Converts a linked list back to an array
 * @param {ListNode} node
 * @returns {Array<number>}
 */
const getLinkedListValues = (node) => {
   const values = [];
   while (node) {
      values.push(node.val);
      node = node.next;
   }
   return values
}


/**
 * Compare two linked lists value by value.
 * @param {ListNode} root1 
 * @param {ListNode} root2 
 * @returns {boolean}
 */
const areLinkedListsEqueal = (root1, root2) => {
   let node1 = root1;
   let node2 = root2;

   while (node1 || node2) {
      if (node1 === null && node2 === null) {
         return true
      } else if (
         (node1 === null || node2 === null) ||
         node1.val !== node2.val
      ) {
         return false
      }
      node1 = node1.next;
      node2 = node2.next;
   }
   return true
}

File: codesite/python_problems/static/python_problems/linked_list_utils.py
from typing import Optional, List


class ListNode:
    """
    Definition for singly-linked list.
    """

    def __init__(self, val=None, next=None):
        self.val = val
        self.next = next


def build_linked_list_deprecated(numbers: List[int]):
    """
    Build linked list from list.
    """
    node = None
    for number in reversed(numbers):
        node = ListNode(number, node)
    return node


def build_linked_list(numbers: List[int], cycle_position: int = None):
    """
    Build linked list with cycle from list.
    """
    anchor = node = ListNode()
    has_cycle = False

    for position, number in enumerate(numbers):
        node.next = ListNode(number)
        node = node.next
        
        if position == cycle_position:
            cycle_node = node
            has_cycle = True

    if has_cycle:
        node.next = cycle_node
    return anchor.next


def get_linked_list_values(root):
    """
    Return linked list values in list.
    """
    values = []
    while root:
        values.append(root.val)
        root = root.next
    return values
